app/src/main/java/nl/nfi/cellscanner/UserDataUploadWorker.java

```diff
@@ -1,106 +1,262 @@
-
-                }
-
-                // disconnect from the server
-                con.logout();
-                con.disconnect();
+            con.enterLocalPassiveMode(); // important!
+            con.setFileType(FTP.BINARY_FILE_TYPE);
+
+            // get the file and send it. A
+            boolean result = con.storeFile(dest_filename, source);
+            source.close();
+
+            if (result) {
+                Log.e("cellscanner", "upload result: succeeded");
+
+                /*when file has been uploaded, the old data can be flushed*/
+
+            } else {
+                Log.e("cellscanner", "upload result: Failed");
             }
+
+            // disconnect from the server
+            con.logout();
+            con.disconnect();
         }
-        catch (Exception e)
-        {
-            e.printStackTrace();
-            ExportResultRepository.storeExportResult(getApplicationContext(), timestamp, false, e.getMessage(), ExportResultRepository.AUTO);
+    }
+
+    private static void uploadSftp(Context ctx, InputStream source, String dest_filename, String host, int port, String user, String password) throws JSchException, SftpException, IOException {
+        JSch jsch = new JSch();
+        jsch.setKnownHosts(new ByteArrayInputStream(ctx.getResources().getText(R.string.ssh_known_hosts).toString().getBytes()));
+        Session session = jsch.getSession(user, host, port == -1 ? 22 : port);
+        if (password != null)
+            session.setPassword(password);
+        session.connect();
+        try {
+            ChannelSftp channel = (ChannelSftp) session.openChannel("sftp");
+            channel.connect();
+            channel.put(source, dest_filename);
+            channel.disconnect();
+        } finally {
+            session.disconnect();
+        }
+    }
+
+    private static void upload(Context ctx, String url_spec) throws URISyntaxException, IOException, SftpException, JSchException {
+        URI uri = new URI(url_spec);
+        String userinfo = uri.getUserInfo();
+        String username = null, password = null;
+        if (userinfo != null) {
+            int sep = userinfo.indexOf(':');
+            if (sep == -1) {
+                username = userinfo;
+                password = null;
+            } else {
+                username = userinfo.substring(0, sep);
+                password = userinfo.substring(sep + 1);
+            }
+        }
+
+        long timestamp = new Date().getTime() / 1000L;
+        String dest_filename = String.format("%s-%d.sqlite3.gz", PreferencesActivity.getInstallID(ctx), timestamp);
+
+        File dbfile = Utils.createTempFile(ctx);
+        try {
+            Utils.copyFileGzipped(Database.getDataFile(ctx), dbfile);
+            InputStream source = new FileInputStream(dbfile);
+            try {
+                if (uri.getScheme() == "ftp")
```

Comment:
I detect that this code is problematic. According to the [Bad practice (BAD_PRACTICE)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#bad-practice-bad-practice), [ES: Comparison of String objects using == or != (ES_COMPARING_STRINGS_WITH_EQ)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#es-comparison-of-string-objects-using-or-es-comparing-strings-with-eq).
This code compares java.lang.String objects for reference equality using the == or != operators. Unless both strings are either constants in a source file, or have been interned using the String.intern() method, the same string value may be represented by two different String objects. Consider using the equals(Object) method instead.